docs/languages/csharp/expression-trees.md

```diff
@@ -0,0 +1,132 @@
+        - Implicit variable type declaration (`var sum`)
+            * Implicit var keyword (`var`)
+            * Variable name declaration (`sum`)
+    * assignment expression (`=`)
+    * Binary Addition Expression (`1 + 2`)
+        - left operand (`1`)
+        - addition operator (`+`)
+        - right operand (`2`)
+
+This may look complicated, but it is very powerful. Following the same process, you can decompose
+much more complicated expressions. Consider this expression:
+```cs
+var finalAnswer = this.SecretSauceFuncion(
+    currentState.createInterimResult(), currentState.createSecondValue(1, 2),
+    decisionServer.considerFinalOptions("hello")) +
+    MoreSecretSauce('A', DateTime.Now, true);
+```
+
+The expression above is also a variable declaration with an assignment.
+In this instance, the right hand side of the assignment is a much more complicated tree.
+I'm not going to decompose this expression, but consider what the different nodes might
+be. There are method calls for the current object, one that has an explicit `this`
+receiver, one that does not. There are method calls using other receiver objects,
+there are constant arguments of different types. And finally, there is a binary
+addition operator. Depending on the return type of `SecretSauceFunction()` or
+`MoreSecretSauce()`, that binary addition operator may be a method call to an
+overridden addition operator.
+
+Despite this perceived complexity, the expression above creates a tree strcuture
+that can be navigated as easily as the first sample. You can keep traversing
+child nodes to find leaf nodes in the expression. Parent nodes will have
+references to their children, and each node has a property that describes
+what kind of node it is.
+
+The structure of an expression tree is very consistent. Once you've learned
+the basics, you can understand even the most complex code. The elegance in
+the data structure explains how the C# compiler can analyze the most complex
+C# programs and create proper output from that complicated source code.
+
+Once you become familiar with the structure of expression trees, you will
+find that knowledge you've gained quickly enables you to work with many
+more and more advanced scenarios. There is incredible power to expression
+trees.
+## Uses for Expression Trees
+You've likely already worked with Expression Trees in libraries that
+you use. One of the most common is Entity Framework (link?). When you
+write queries against Entity Framework sets, you define expression trees.
+The code inside Entity Framework traverses those expression trees, and
+translates the C# code that you wrote into parameterized SQL queries.
+Regardless of the complexity of the query expressions you create, whether
+you use the query syntax, or the fluent method call syntax, Entity Framework
+translates the expression tree into SQL statements.
+
+There are some limitations to what Entity Framework can translate. In general,
```

Comment:
I think this paragraph is a little too much detail, as it gets into EF details rather than expression trees.  Might be good to just provide a link to EF docs and ask the reader to look there for more details.